flink-runtime/src/main/java/org/apache/flink/runtime/history/FsJobArchivist.java

```diff
@@ -112,12 +112,19 @@ public static Path archiveJob(Path rootPath, JobID jobId, Collection<ArchivedJso
 			JsonNode archive = mapper.readTree(output.toByteArray());
 
 			Collection<ArchivedJson> archives = new ArrayList<>();
-			for (JsonNode archivePart : archive.get(ARCHIVE)) {
-				String path = archivePart.get(PATH).asText();
-				String json = archivePart.get(JSON).asText();
+			for (JsonNode archivePart : failOnNull(archive.get(ARCHIVE))) {
+				String path = failOnNull(archivePart.get(PATH)).asText();
+				String json = failOnNull(archivePart.get(JSON)).asText();
 				archives.add(new ArchivedJson(path, json));
 			}
 			return archives;
 		}
 	}
+
+	private static <X> X failOnNull(X value) throws IOException {
+		if (value == null) {
+			throw new IOException("Job archive did not conform to expected format.");
```

Comment:
Maybe we could add which `file` contained the invalid format.

Reply:
I've reworked this commit (now uses a simple try-catch block) and include the filename in the error message.